### PR TITLE
fix #8212(get name of procedure)

### LINKED ISSUE #8212: we need a standard way to get name of procedure
Is https://stackoverflow.com/a/51198228/1426932 the best we can do to get access (at compile time) to name of procedure in a macro?

should that information be in instantiationInfo instead or would that incur some cost (bloat, compile time, memory)?

reposting here:
```nim
template procName*: string =
  var name_D20180705T114826: cstring
  {.emit: "`name_D20180705T114826` = __func__;".}
  let temp_D20180705T114826=($name_D20180705T114826).rsplit('_', 2)
    #CHECKME: IMPROVE; the magic '4' chosen to be enough for most cases
    # EG: bar_baz_9c8JPzPvtM9azO6OB23bjc3Q_3
  if temp_D20180705T114826.len>=3 and temp_D20180705T114826[2].len < 4:
    ($name_D20180705T114826).rsplit('_', 2)[0]
  else:
    # EG: foo_9c8JPzPvtM9azO6OB23bjc3Q
    ($name_D20180705T114826).rsplit('_', 1)[0]
```

### PR DESCRIPTION
**Pull request by ringabout**
fix #8212

TODO
- [ ] more tests
- [ ] docs
- [ ] think of where the magics should be put
- [ ] ...

### CODE REVIEW

File: compiler/semexprs.nim [hunk 1]
```diff
@@ -1894,6 +1894,16 @@ proc semDefined(c: PContext, n: PNode): PNode =
   result.info = n.info
   result.typ = getSysType(c.graph, n.info, tyBool)
 
+proc semGetProcname(c: PContext, n: PNode): PNode =
+  let sym = getCurrOwner(c)
+  case sym.kind
+  of skModule:
+     localError(c.config, n.info, "'getProcname' cannot be used in the top level")
```
Comment (Araq) [inline compiler/semexprs.nim]:
```suggestion
   localError(c.config, n.info, "'getProcname' cannot be used in the top level")
```

File: compiler/semexprs.nim [hunk 2]
```diff
@@ -1894,6 +1894,16 @@ proc semDefined(c: PContext, n: PNode): PNode =
   result.info = n.info
   result.typ = getSysType(c.graph, n.info, tyBool)
 
+proc semGetProcname(c: PContext, n: PNode): PNode =
+  let sym = getCurrOwner(c)
+  case sym.kind
+  of skModule:
+   localError(c.config, n.info, "'getProcname' cannot be used in the top level")
```
Comment (ringabout) [inline compiler/semexprs.nim]:
```suggestion
    localError(c.config, n.info, "'getProcname' cannot be used in the top level")
```

File: lib/std/magics.nim [hunk 3]
```diff
@@ -0,0 +1,3 @@
+proc getProcname*(): string {.magic: "GetProcname".} =
+  ## Returns the names of procs/macros. It cannot be used
```
Comment (ringabout) [inline lib/std/magics.nim]:
```suggestion
  ## Returns the name of a proc or a macro. It cannot be used
```

**Comment by timotheecour** [comment]
@xflywind see instead https://github.com/nim-lang/Nim/pull/17641, I've been using this code for a while, it's very helpful in particular for improving debugging, logging etc. I finally made a PR for it.

**Comment by konsumlamm** [comment]
If this doesn't get closed by #17641:
`magics` is very undescriptive, it doesn't tell anything about what the module does (and it's not even the only module using compiler magic). Besides, why make a new module for this instead of adding it to `typetraits`?

**Comment by ringabout** [comment]
it is not related to type, i think reflection is a better name. this PR will be closed after timothee's PR is merged

**Comment by arnetheduck** [comment]
this feels a bit limited / constrained as far as api go? most of the time, one wants proc names not only of the current proc but parents, and due to global namespacing in nim, the module name as well (and other disambiguators like parameters and their names - I imagine a more complete api would allow more reflection where the current proc name is just a small part of it)

**Comment by arnetheduck** [comment]
https://docs.microsoft.com/en-us/dotnet/api/system.reflection.methodbase.getcurrentmethod?view=net-5.0 could be a good starting point

**Comment by ringabout** [comment]
@arnetheduck
see https://github.com/nim-lang/Nim/pull/17641 for  a better solution.

**Pull request closed without merging**